modules/renter/download.go

```diff
@@ -1,182 +1,290 @@
-	data, err := hf.downloader.Sector(p.MerkleRoot)
-	if err != nil {
-		return nil, err
-	}
+// newDownload initializes and returns a download object.
+func (f *file) newDownload(destination string) *download {
+	d := &download{
+		atomicChunkIndex: 0,
+		atomicReceived:   0,
 
-	// generate decryption key
-	key := deriveKey(hf.masterKey, p.Chunk, p.Piece)
+		erasureCode: f.erasureCode,
+		chunkSize:   f.chunkSize(),
+		fileSize:    f.size,
 
-	// decrypt and return
-	return key.DecryptBytes(data)
+		startTime:   time.Now(),
+		siapath:     f.name,
+		destination: destination,
+
+		downloadFinished: make(chan error),
+	}
+	return d
 }
 
-// newHostFetcher creates a new hostFetcher.
-func newHostFetcher(d contractor.Downloader, pieces []pieceData, masterKey crypto.TwofishKey) *hostFetcher {
-	// make piece map
-	pieceMap := make(map[uint64][]pieceData)
-	for _, p := range pieces {
-		pieceMap[p.Chunk] = append(pieceMap[p.Chunk], p)
+// managedDownloadIteraiton downloads a chunk from the next available file.
+func (r *Renter) managedDownloadIteration() {
+	// Get the set of available workers.
+	availableWorkers := make([]*worker, 0)
+	id := r.mu.RLock()
+	for _, worker := range r.workerPool {
+		availableWorkers = append(availableWorkers, worker)
 	}
-	return &hostFetcher{
-		downloader: d,
-		pieceMap:   pieceMap,
-		masterKey:  masterKey,
+	r.mu.RUnlock(id)
+
+	// Grab the next file to be downloaded.
+	id = r.mu.RLock()
+	if len(r.downloadQueue) == 0 {
+		// TODO: Block until either there's another download, or until
+		// shutdown, or until enough time has passed that the workers should be
+		// reassembled?
+		//
+		// TODO: This should probably be moved to the outer loop, and then the
+		// managedDownloadIteraton can be turned into managedDownloadFile
+		// taking a file as input.
+		r.mu.RUnlock(id)
+		return
 	}
-}
+	nextDownload := r.downloadQueue[0]
+	r.downloadQueue = r.downloadQueue[1:]
+	r.mu.RUnlock(id)
 
-// checkHosts checks that a set of hosts is sufficient to download a file.
-func checkHosts(hosts []fetcher, minPieces int, numChunks uint64) error {
-	for i := uint64(0); i < numChunks; i++ {
-		pieces := 0
-		for _, h := range hosts {
-			pieces += len(h.pieces(i))
-		}
-		if pieces < minPieces {
-			return errInsufficientHosts
+	// Reject this download if there are not enough workers to complete the
+	// download.
+	if len(availableWorkers) < nextDownload.erasureCode.MinPieces() {
+		// Signal that the download has errored and quit.
+		nextDownload.downloadFinished <- errInsufficientHosts
+		return
+	}
+
+	// TODO: Exclude the most expensive workers, and perhaps the slowest as
+	// well.
+
+	// Open a file handle for the download.
+	fileDest, err := os.OpenFile(nextDownload.destination, os.O_CREATE|os.O_RDWR|os.O_TRUNC, 0600)
+	if err != nil {
+		nextDownload.downloadFinished <- err
+		return
+	}
+
+	// Grab each chunk from the download.
+	chunkDownloads := make([]chunkDownload, nextDownload.fileSize/nextDownload.chunkSize)
```

Comment:
there should be a `file.numChunks()` method for calculating this. Simple division will produce the wrong value unless the file size is an even multiple of the chunk size.